classes/requests/psr/class-truelayer-http-client.php

```diff
@@ -0,0 +1,162 @@
+use KrokedilTrueLayerDeps\Psr\Http\Message\RequestInterface;
+use KrokedilTrueLayerDeps\Psr\Http\Message\ResponseInterface;
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Class TrueLayer_Http_Client
+ */
+class TrueLayer_Http_Client implements ClientInterface {
+	/**
+	 * The log title to use for logging.
+	 *
+	 * @var string
+	 */
+	protected $log_title;
+
+	/**
+	 * The payment ID to use for the logging.
+	 *
+	 * @var string|null
+	 */
+	protected $payment_id;
+
+	/**
+	 * Class constructor.
+	 *
+	 * @param string      $log_title The log title to use for logging.
+	 * @param string|null $payment_id The payment ID to use for the logging.
+	 *
+	 * @return void
+	 */
+	public function __construct( $log_title = 'TrueLayer Request', $payment_id = null ) {
+		$this->log_title  = $log_title;
+		$this->payment_id = $payment_id;
+	}
+
+	/**
+	 * Sends a PSR-7 request and returns a PSR-7 response.
+	 *
+	 * @param RequestInterface $request The request to send.
+	 *
+	 * @return ResponseInterface
+	 */
+	public function sendRequest( RequestInterface $request ): ResponseInterface {
+		$uri          = $request->getUri();
+		$args         = $this->formatArgs( $request );
+		$http_version = $request->getProtocolVersion();
+
+		// Make the request.
+		$response = wp_remote_request( $uri, $args );
+
+		// Create the PSR-7 response.
+		$response_code    = wp_remote_retrieve_response_code( $response );
+		$response_headers = wp_remote_retrieve_headers( $response );
+		$response_reason  = wp_remote_retrieve_response_message( $response );
+		$response_body    = wp_remote_retrieve_body( $response );
+
+		// Ensure the response code is numeric.
+		$response_code = is_numeric( $response_code ) ? (int) $response_code : 400;
+
+		// Format the headers to ensure they are an array.
+		$response_headers = is_array( $response_headers ) ? $response_headers : iterator_to_array( $response_headers );
+
+		// Create the PSR-7 response.
+		$response = new Response( $response_code, $response_headers, $response_body, $http_version, $response_reason );
+
+		$this->logRequest( $uri, $args, $response_body, $response_code );
+
+		return $response;
+	}
+
+	/**
+	 * Format the arguments for the WordPress HTTP API request.
+	 *
+	 * @param RequestInterface $request The request to send.
+	 *
+	 * @return array
+	 */
+	private function formatArgs( RequestInterface $request ): array {
+		$args = array(
+			'method'      => $request->getMethod(),
+			'headers'     => $this->formatHeaders( $request ),
+			'httpversion' => $request->getProtocolVersion(),
+			'user-agent'  => $this->getUserAgent(),
+			'timeout'     => apply_filters( 'truelayer_request_timeout', 10 ),
+		);
+
+		// Add the request body if it is set.
+		if ( $request->getBody()->getSize() > 0 ) {
+			$body = json_decode( (string) $request->getBody(), true );
+			if ( ! empty( $body ) ) {
+				$args['body'] = apply_filters( 'truelayer_request_args', (string) wp_json_encode( $body ) );
```

Comment:
Lets also remove the body from here, unless you want to  manually remove any sensitive data from it, such as all the user information.

Reply:
Sorry but do you mean from the filter or the args passed to the request later?

The filter is there to give merchants the option to modify the data passed in the request if needed. Is that not something we want to support going forward?